core/ledger/kvledger/benchmark/chainmgmt/chains.go

```diff
@@ -0,0 +1,153 @@
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package chainmgmt
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/common/configtx/test"
+	"github.com/hyperledger/fabric/core/ledger"
+	"github.com/hyperledger/fabric/core/ledger/ledgermgmt"
+	"github.com/hyperledger/fabric/core/ledger/ledgermgmt/ledgermgmttest"
+	"github.com/hyperledger/fabric/core/ledger/util/couchdb"
+)
+
+// ChainID is a type used for the ids for the chains for experiments
+type ChainID int
+
+func (chainID ChainID) String() string {
+	return fmt.Sprintf("%s%04d", "chain-", chainID)
+}
+
+// SimulationResult is a type used for simulation results
+type SimulationResult []byte
+
+// chainsMgr manages chains for experiments
+type chainsMgr struct {
+	ledgerMgr *ledgermgmt.LedgerMgr
+	mgrConf   *ChainMgrConf
+	batchConf *BatchConf
+	initOp    chainInitOp
+	chainsMap map[ChainID]*Chain
+	wg        *sync.WaitGroup
+}
+
+func newChainsMgr(mgrConf *ChainMgrConf, batchConf *BatchConf, initOp chainInitOp) *chainsMgr {
+	dataDir := filepath.Join(mgrConf.DataDir, "ledgersData")
+	ledgermgmtInitializer := ledgermgmttest.NewInitializer(dataDir)
+	ledgermgmtInitializer.Config.HistoryDBConfig.Enabled = true
+	if os.Getenv("useCouchDB") == "true" {
+		ledgermgmtInitializer.Config.StateDBConfig.StateDatabase = "CouchDB"
+		ledgermgmtInitializer.Config.StateDBConfig.CouchDB = &couchdb.Config{
+			Address:            "localhost:5984",
+			RedoLogPath:        filepath.Join(dataDir, "couchdbRedologs"),
+			UserCacheSizeMBs:   500,
+			MaxBatchUpdateSize: 500,
+		}
+	}
+	ledgerMgr := ledgermgmt.NewLedgerMgr(ledgermgmtInitializer)
+	return &chainsMgr{ledgerMgr, mgrConf, batchConf, initOp, make(map[ChainID]*Chain), &sync.WaitGroup{}}
+}
+
+func (m *chainsMgr) createOrOpenChains() []*Chain {
+	numChains := m.mgrConf.NumChains
+	switch m.initOp {
+	case ChainInitOpCreate:
+		for i := 0; i < numChains; i++ {
+			chainID := ChainID(i)
+			ledgerID := chainID.String()
+			gb, err := test.MakeGenesisBlock(ledgerID)
+			panicOnError(err)
+			peerLedger, err := m.ledgerMgr.CreateLedger(ledgerID, gb)
+			panicOnError(err)
+			c := newChain(chainID, peerLedger, m)
+			m.chainsMap[chainID] = c
+		}
+
+	case ChainInitOpOpen:
+		for i := 0; i < numChains; i++ {
+			chainID := ChainID(i)
+			peerLedger, err := m.ledgerMgr.OpenLedger(chainID.String())
+			panicOnError(err)
+			c := newChain(chainID, peerLedger, m)
+			m.chainsMap[chainID] = c
+		}
+
+	default:
+		panic(fmt.Errorf("unknown chain init opeartion"))
+	}
+	return m.chains()
+}
+
+func (m *chainsMgr) chains() []*Chain {
+	chains := []*Chain{}
+	for _, chain := range m.chainsMap {
+		chains = append(chains, chain)
+	}
+	return chains
+}
+
+func (m *chainsMgr) waitForChainsToExhaustAllBlocks() {
+	m.wg.Wait()
+	m.ledgerMgr.Close()
+}
+
+// Chain extends ledger.PeerLedger and the experiments invoke ledger functions via a chain type
```

Comment:
```suggestion
// Chain embeds ledger.PeerLedger and the experiments invoke ledger functions via a chain type
```
There is no inheritance in `go`.